src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WinRTWebSocket.cs

```diff
@@ -515,5 +549,80 @@ private void UpdateStateCloseReceived()
             }
         }
         #endregion
+
+        #region Helpers
+        private static bool InitMessageWebSocketClientCertificateSupported()
+        {
+            return ApiInformation.IsPropertyPresent(
+                "Windows.Networking.Sockets.MessageWebSocketControl",
+                "ClientCertificate");
+        }
+
+        private static X509Certificate2 GetEligibleClientCertificate(X509Certificate2Collection candidateCerts)
+        {
+            Debug.Assert(candidateCerts.Count > 0);
+
+            // Reduce the set of certificates to match the proper 'Client Authentication' criteria.
+            candidateCerts = candidateCerts.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, false);
+            candidateCerts = candidateCerts.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, false);
+
+            // Build a new collection with certs that have a private key. We need to do this manually because there is
+            // no X509FindType to match this criteria.
+            var eligibleCerts = new X509Certificate2Collection();
+            foreach (X509Certificate2 cert in candidateCerts)
+            {
+                if (cert.HasPrivateKey)
+                {
+                    eligibleCerts.Add(cert);
+                }
+            }
+
+            if (eligibleCerts.Count > 0)
```

Comment:
Depending on your caring levels this function is anywhere from "meh" to "eep!"  because Find returns a collection of clones instead of a filtered collection.

So let's say that someone adds in their entire CU\My store, and that it happens to be 45 certificates. 5 are public-key only (one of which has no EKU extension).  All 45 have DigitalSignature.  And 5 have either the client EKU or no EKU extension.

* 45 certs come in, which were the user's responsibility (and unless the user's me, they probably won't ever dispose them)
* Find (DigitalSignature) returns 45 new certificates.
* Find (ClientAuthOid) returns 5 new certificates.
* HasPrivateKey filtering reduces that to 4 eligible
* Why does web socket not have a server trusted issuers list filter?  Oh well.
* One cert, guaranteed different than an input cert, is returned.

So this function created 50 new finalizable objects, returned one of them, and let 49 of them go to the finalizer queue.

Best practices for using `Find`:

* Don't call it if you don't need to.  (if count == 1 => just return it)
* Make sure that you're reducing the candidate space as fast as possible.
  * In this case, client EKU is probably more rare than the DigitalSignature KU.
  * Though your input set it small, so maybe it doesn't matter.
* Once you've called Find you own the returned objects.  Dispose them as able.

Using my made up numbers from above, if you filtered out the !HasPrivateKey certs first you'd be down to 40 certs into the first call to Find.  If that was the client EKU test you'd now be down to 4 new certs, and 4 more new certs verifying the KU, one gets returned.  7 finalizations instead of 49.  (not to mention fewer allocations)